## Imports

In [6]:
import os
import requests
import pandas as pd

## Build the financial dataset

In [2]:
financial_data = {
    "year": [2019, 2020, 2021, 2022, 2023, 2024, 2025],
    "revenue_gbp_m": [163, 220, 636, 850, 1700, 3100, 4500],
    "net_profit_gbp_m": [-107, -206, 26, None, 438, 790, 1300],
    "users_millions": [10, 14.5, 16, 27, 38, 52.5, 68],
    "valuation_usd_bn": [1.7, 5.5, 33, 33, 33, 45, 75],
}

df_financials = pd.DataFrame(financial_data)
df_financials

,year,revenue_gbp_m,net_profit_gbp_m,users_millions,valuation_usd_bn
0,2019,163,-107.0,10.0,1.7
1,2020,220,-206.0,14.5,5.5
2,2021,636,26.0,16.0,33.0
3,2022,850,NaN,27.0,33.0
4,2023,1700,438.0,38.0,33.0
5,2024,3100,790.0,52.5,45.0
6,2025,4500,1300.0,68.0,75.0


In [3]:
financial_data = {
    "year": [2019, 2020, 2021, 2022, 2023, 2024, 2025],
    "revenue_gbp_m": [163, 222, 636, 850, 1700, 3100, 4500],
    "net_profit_gbp_m": [-107, -206, 26, None, 438, 790, 1300],
    "profit_before_tax_gbp_m": [None, None, None, None, 438, 1089, 1700],
    "users_millions": [10, 14.5, 16, 27, 38, 52.5, 68],
    "business_customers_k": [220, 500, None, 500, None, None, 767],
    "valuation_usd_bn": [1.7, 5.5, 33, 33, 33, 45, 75],
    "customer_balances_gbp_bn": [None, None, None, None, 18.2, 30.2, 50.2],
    "monthly_transactions_m": [None, None, None, None, 590, 940, None],
    "subscription_revenue_gbp_m": [None, None, None, None, 244, 423, 708],
    "arpu_usd": [None, None, None, 45, 59, None, None],
}

df_financials = pd.DataFrame(financial_data)
df_financials

,year,revenue_gbp_m,net_profit_gbp_m,profit_before_tax_gbp_m,users_millions,business_customers_k,valuation_usd_bn,customer_balances_gbp_bn,monthly_transactions_m,subscription_revenue_gbp_m,arpu_usd
0,2019,163,-107.0,NaN,10.0,220.0,1.7,NaN,NaN,NaN,NaN
1,2020,222,-206.0,NaN,14.5,500.0,5.5,NaN,NaN,NaN,NaN
2,2021,636,26.0,NaN,16.0,NaN,33.0,NaN,NaN,NaN,NaN
3,2022,850,NaN,NaN,27.0,500.0,33.0,NaN,NaN,NaN,45.0
4,2023,1700,438.0,438.0,38.0,NaN,33.0,18.2,590.0,244.0,59.0
5,2024,3100,790.0,1089.0,52.5,NaN,45.0,30.2,940.0,423.0,NaN
6,2025,4500,1300.0,1700.0,68.0,767.0,75.0,50.2,NaN,708.0,NaN


## Derived insight metrics

In [4]:
df_financials["revenue_growth_pct"] = df_financials["revenue_gbp_m"].pct_change() * 100
df_financials["users_growth_pct"] = df_financials["users_millions"].pct_change() * 100
df_financials["profit_margin_pct"] = (df_financials["net_profit_gbp_m"] / df_financials["revenue_gbp_m"]) * 100
df_financials["revenue_per_user_gbp"] = (df_financials["revenue_gbp_m"] * 1_000_000) / (df_financials["users_millions"] * 1_000_000)
df_financials["balances_per_user_gbp"] = (df_financials["customer_balances_gbp_bn"] * 1000) / df_financials["users_millions"]
df_financials["subscription_share_pct"] = (df_financials["subscription_revenue_gbp_m"] / df_financials["revenue_gbp_m"]) * 100

df_financials

,year,revenue_gbp_m,net_profit_gbp_m,profit_before_tax_gbp_m,users_millions,business_customers_k,valuation_usd_bn,customer_balances_gbp_bn,monthly_transactions_m,subscription_revenue_gbp_m,arpu_usd,revenue_growth_pct,users_growth_pct,profit_margin_pct,revenue_per_user_gbp,balances_per_user_gbp,subscription_share_pct
0,2019,163,-107.0,NaN,10.0,220.0,1.7,NaN,NaN,NaN,NaN,NaN,NaN,-65.644172,16.300000,NaN,NaN
1,2020,222,-206.0,NaN,14.5,500.0,5.5,NaN,NaN,NaN,NaN,36.196319,45.000000,-92.792793,15.310345,NaN,NaN
2,2021,636,26.0,NaN,16.0,NaN,33.0,NaN,NaN,NaN,NaN,186.486486,10.344828,4.088050,39.750000,NaN,NaN
3,2022,850,NaN,NaN,27.0,500.0,33.0,NaN,NaN,NaN,45.0,33.647799,68.750000,NaN,31.481481,NaN,NaN
4,2023,1700,438.0,438.0,38.0,NaN,33.0,18.2,590.0,244.0,59.0,100.000000,40.740741,25.764706,44.736842,478.947368,14.352941
5,2024,3100,790.0,1089.0,52.5,NaN,45.0,30.2,940.0,423.0,NaN,82.352941,38.157895,25.483871,59.047619,575.238095,13.645161
6,2025,4500,1300.0,1700.0,68.0,767.0,75.0,50.2,NaN,708.0,NaN,45.161290,29.523810,28.888889,66.176471,738.235294,15.733333


In [12]:
df_financials.to_csv("E:/project/Revolut-analysis-deashboard/Data/revolut_financials_clean.csv", index=False)
print("E:/project/Revolut-analysis-deashboard/Data/revolut_financials_clean.csv")

E:/project/Revolut-analysis-deashboard/Data/revolut_financials_clean.csv


##  Fetch filing history(Company House) 

In [7]:
API_KEY = "c139ab29-2252-47a0-99fb-ee79c316ba53"
COMPANY_NUMBER = "12743269"  # Revolut Group Holdings Ltd

url = f"https://api.company-information.service.gov.uk/company/{COMPANY_NUMBER}/filing-history"
response = requests.get(url, auth=(API_KEY, ""))

if response.status_code == 200:
    data = response.json()
    filings = pd.json_normalize(data["items"])
    print(f"Fetched {len(filings)} filings")
    filings.head()
else:
    print(f"Error: {response.status_code}")
    print(response.text)

Fetched 25 filings


In [8]:
accounts_filings = filings[filings["type"] == "AA"][["date", "description", "links.document_metadata"]]
accounts_filings

,date,description,links.document_metadata
4,2026-04-03,accounts-with-accounts-type-group,https://document-api.company-information.servi...


In [9]:
url = f"https://api.company-information.service.gov.uk/company/{12743269}/filing-history"
params = {"category": "accounts", "items_per_page": 50}
response = requests.get(url, auth=(API_KEY, "c139ab29-2252-47a0-99fb-ee79c316ba53"), params=params)

if response.status_code == 200:
    data = response.json()
    filings = pd.json_normalize(data["items"])
    print(f"Fetched {len(filings)} accounts filings")
    filings[["date", "description", "links.document_metadata"]]
else:
    print(f"Error: {response.status_code}")
    print(response.text)

Fetched 7 accounts filings


In [15]:
doc_metadata_url = accounts_filings.iloc[0]["links.document_metadata"]

# Get the actual document link
doc_response = requests.get(doc_metadata_url, auth=(API_KEY, ""))
doc_data = doc_response.json()

pdf_url = doc_data["links"]["document"]
pdf_response = requests.get(pdf_url, auth=(API_KEY, ""), headers={"Accept": "application/pdf"})

with open("E:/project/Revolut-analysis-deashboard/Data/revolut_accounts_2025.pdf", "wb") as f:
    f.write(pdf_response.content)

print("Downloaded accounts PDF")

Downloaded accounts PDF


In [16]:
doc_metadata_url = accounts_filings.iloc[0]["links.document_metadata"]
print(f"Requesting: {doc_metadata_url}")

doc_response = requests.get(doc_metadata_url, auth=(API_KEY, ""))

print(f"Status code: {doc_response.status_code}")
print(f"Response text (first 500 chars): {doc_response.text[:500]}")

Requesting: https://document-api.company-information.service.gov.uk/document/2E1FFIBSA-oxaxZIVT-JK1rRpl0WtfornqYq1gyYm70
Status code: 200
Response text (first 500 chars): {"company_number":"12743269","barcode":"AEYL33H6","significant_date":"2025-12-31T00:00:00Z","significant_date_type":"made-up-date","category":"accounts","pages":207,"filename":"","created_at":"2026-04-07T23:00:55.493861388Z","etag":"","links":{"self":"https://document-api.company-information.service.gov.uk/document/2E1FFIBSA-oxaxZIVT-JK1rRpl0WtfornqYq1gyYm70","document":"https://document-api.company-information.service.gov.uk/document/2E1FFIBSA-oxaxZIVT-JK1rRpl0WtfornqYq1gyYm70/content"},"resour


In [17]:
print(doc_data["links"]["document"])

https://document-api.company-information.service.gov.uk/document/2E1FFIBSA-oxaxZIVT-JK1rRpl0WtfornqYq1gyYm70/content


In [18]:
file_path = "E:/project/Revolut-analysis-deashboard/Data/revolut_accounts_2025.pdf"
size_kb = os.path.getsize(file_path) / 1024
print(f"File size: {size_kb:.1f} KB")

File size: 11401.3 KB


In [19]:
document_url = doc_data["links"]["document"]

pdf_response = requests.get(
    document_url,
    auth=(API_KEY, ""),
    headers={"Accept": "application/pdf"}
)

print(f"Status code: {pdf_response.status_code}")
print(f"Content-Type: {pdf_response.headers.get('Content-Type')}")
print(f"Content length: {len(pdf_response.content)} bytes")

Status code: 200
Content-Type: application/pdf
Content length: 11674890 bytes


## Fetching Revolut's Annual Report from Companies House

**Goal:** Download Revolut Group Holdings Ltd's official annual accounts (primary source, not press coverage).

**Process:**
1. `GET /company/{number}/filing-history?category=accounts` → list of accounts filings (dates + metadata links)
2. `GET {metadata_link}` → info *about* the document (pages, date, and a `document` URL — not the file itself)
3. `GET {document_url}` with header `Accept: application/pdf` → the actual PDF bytes
4. Save bytes to `.pdf` with `open(path, "wb")`

**Auth:** API key goes in `auth=(API_KEY, "")` — key as username, empty password.

**Result:** `revolut_accounts_2025.pdf` (~11.4 MB, 207 pages) — Revolut's real FY2025 annual report, filed with UK Companies House (company no. 12743269).

**Next step:** extract revenue, profit, expenses, and customer balance figures from the financial statements section of the PDF.

##  Install a PDF reading library

In [20]:
pip install pdfplumber

Defaulting to user installation because normal site-packages is not writeable
   ---------------------------------------- 0.0/6.6 MB ? eta -:--:--
   ---- ----------------------------------- 0.8/6.6 MB 5.4 MB/s eta 0:00:02
   --------- ------------------------------ 1.6/6.6 MB 4.7 MB/s eta 0:00:02
   --------- ------------------------------ 1.6/6.6 MB 4.7 MB/s eta 0:00:02
   ------------ --------------------------- 2.1/6.6 MB 2.6 MB/s eta 0:00:02
   ----------------- ---------------------- 2.9/6.6 MB 2.9 MB/s eta 0:00:02
   ----------------------- ---------------- 3.9/6.6 MB 3.1 MB/s eta 0:00:01
   ---------------------------- ----------- 4.7/6.6 MB 3.3 MB/s eta 0:00:01
   ------------------------------- -------- 5.2/6.6 MB 3.1 MB/s eta 0:00:01
   ------------------------------------ --- 6.0/6.6 MB 3.3 MB/s eta 0:00:01
   ---------------------------------------- 6.6/6.6 MB 3.3 MB/s  0:00:01
   ---------------------------------------- 0.0/3.9 MB ? eta -:--:--
   -------- ---------------

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
